src/core/agent/services/keriaNotificationService.ts

```diff
@@ -0,0 +1,743 @@
+          await new Promise((rs) =>
+            setTimeout(rs, KeriaNotificationService.CHECK_READINESS_INTERNAL)
+          );
+          continue;
+        }
+
+        const startFetchingIndex =
+          notificationQuery.nextIndex > 0 ? notificationQuery.nextIndex - 1 : 0;
+
+        let notifications;
+        try {
+          notifications = await this.props.signifyClient
+            .notifications()
+            .list(startFetchingIndex, startFetchingIndex + 24);
+        } catch (error) {
+          const errorMessage = (error as Error).message;
+          /** If the error is failed to fetch with signify,
+           * we retry until the connection is secured*/
+          if (
+            (/Failed to fetch/gi.test(errorMessage) ||
+              /Load failed/gi.test(errorMessage)) &&
+            this.getKeriaOnlineStatus()
+          ) {
+            // Possible that bootAndConnect is called from @OnlineOnly in between loops,
+            // so check if its gone down to avoid having 2 bootAndConnect loops
+            this.markAgentStatus(false);
+            // This will hang the loop until the connection is secured again
+            await this.connect();
+          } else {
+            throw error;
+          }
+        }
+        if (!notifications) {
+          // KERIA went down while querying, now back online
+          continue;
+        }
+        if (
+          notificationQuery.nextIndex > 0 &&
+          (notifications.notes.length == 0 ||
+            notifications.notes[0].i !== notificationQuery.lastNotificationId)
+        ) {
+          // This is to verify no notifications were deleted for some reason (which affects the batch range)
+          notificationQuery = {
+            nextIndex: 0,
+            lastNotificationId: "",
+          };
+          await this.basicStorage.createOrUpdateBasicRecord(
+            new BasicRecord({
+              id: MiscRecordId.KERIA_NOTIFICATION_MARKER,
+              content: notificationQuery,
+            })
+          );
+          continue;
+        }
+        if (notificationQuery.nextIndex > 0) {
+          // Since the first item is the (next index - 1), we can ignore it
+          notifications.notes.shift();
+        }
+        for (const notif of notifications.notes) {
+          try {
+            await this.processNotification(notif, callback);
+            const nextNotificationIndex = notificationQuery.nextIndex + 1;
+            notificationQuery = {
+              nextIndex: nextNotificationIndex,
+              lastNotificationId: notif.i,
+            };
+            await this.basicStorage.createOrUpdateBasicRecord(
+              new BasicRecord({
+                id: MiscRecordId.KERIA_NOTIFICATION_MARKER,
+                content: notificationQuery,
+              })
+            );
+          } catch (error) {
+            /* eslint-disable no-console */
+            console.error("Error when process a notification", error);
```

Comment:
Can we add: `// @TODO - Consider how to track/retry old notifications we couldn't process`